### PR TITLE
Stitch selection methods

### PR DESCRIPTION
**Pull request by kaalleen**
Builds upon the meander images branch.

Once we talked about making it easier to switch between zigzag and running stitches.
This is my attempt to do that. It updates the dashes as stitch methods are changed.

### CODE REVIEW

**Comment by lexelby** [comment]
Aw, I liked the satin_method change.  Why did you revert that?  I like the idea of treating it as similar to stroke method and fill method.  I'm interested to see what other kinds of satin methods you might come up with :D

I'm really happy you got rid of the 0.5 running stitch thing.  It's about time we punted that, it's from waaaay back in time!

I like how you're trying to standardize on the stroke method.  I think this is a good way to go and makes a lot of sense.  I"m not sure about the dash stuff though.  Setting running stitch in Params will convert the path to dashed.  Removing the dashes _won't_ turn it back to zig-zag.  Every time they run any extension (well, most of them), it'll convert the path back to dashes.  Same for if they set the path to zig-zag in Params, it'll get rid of their dashes and ignore if they set the path dashed.  But ripple doesn't do anything like that, so if the user sets the path to dashed and then sets it as ripple in Params, it'll stay ripple even though the path is dashed.

I'm not sure what I actually want to have happen.  I feel like the dashes should always win, but then again I like the idea that if they choose a dashed line and then set it to zig-zag or ripple, the dashes get removed.  That makes sense.

I kind of feel like we should try to honor whatever change the user makes.  If they select something new in the drop-down, we should apply that to the path.  So if they select running stitch, we should make the path dashed, and if they select one of the others, we should remove the dashes.  But then if they change the dashes in Inkscape, we should respect that and update the stroke_method the next time Ink/Stitch runs.  In other words, I feel like the user's last action should always win, whatever they do.  It won't be easy to code that though...

What do you think?

**Comment by kaalleen** [comment]
> Aw, I liked the satin_method change. Why did you revert that? I like the idea of treating it as similar to stroke method and fill method. I'm interested to see what other kinds of satin methods you might come up with :D

:D I like the idea too. But it seemed to be more complicated than I thought it would be. So I removed it. But can try again.

> I'm really happy you got rid of the 0.5 running stitch thing. It's about time we punted that, it's from waaaay back in time!

Troubleshoot will warn about it, so I thougth it would be fine to remove it. But it actually reminds people to switch the stroke style to dashes.

> I like how you're trying to standardize on the stroke method. I think this is a good way to go and makes a lot of sense. I"m not sure about the dash stuff though. Setting running stitch in Params will convert the path to dashed. Removing the dashes won't turn it back to zig-zag. Every time they run any extension (well, most of them), it'll convert the path back to dashes. Same for if they set the path to zig-zag in Params, it'll get rid of their dashes and ignore if they set the path dashed. But ripple doesn't do anything like that, so if the user sets the path to dashed and then sets it as ripple in Params, it'll stay ripple even though the path is dashed.

I was wondering about it too. Each style would need a clear definition if it has dashes or no dashes.

> What do you think?

To be honest I'm not a fan of the dashes at all. I think it complicates things so much. I'd like to allow everybody to set their stroke style to their own wishes and chose the actual stitch type in the param. Because people miss to set their dashes they have zigzag lines everywhere and they just look bad around corners. I'd prepfer if a stroke always defaults to a running stitch.

**Comment by lexelby** [comment]
> :D I like the idea too. But it seemed to be more complicated than I thought it would be. So I removed it. But can try again.

Honestly, I felt the complexity wasn't too bad, especially when you compare it to the advantages we get in consistency and future expansion.  Let's give it another go :)

> To be honest I'm not a fan of the dashes at all. I think it complicates things so much. I'd like to allow everybody to set their stroke style to their own wishes and chose the actual stitch type in the param. Because people miss to set their dashes they have zigzag lines everywhere and they just look bad around corners. I'd prepfer if a stroke always defaults to a running stitch.

Yeah, I know you've hated the dashes for a long time.  Have you heard complaints from users that the dashes are confusing or cause problems?

I'm not a fan of simple satin at all.  I kind of wish we could get rid of it.  I think it's a terrible default and it's super-confusing for users.  So I'd be a fan of defaulting to running stitch if we could figure out how to make it make sense.

How would we do this?  If someone has an old design with simple satin, we're suddenly going to turn it into running stitch.  That's not great.  We've just broken their design.

We could add a new extension, "Convert old SVGs" that people should run once on all of their SVGs to apply the right stroke_method.  We could make it easy by processing a folder recursively for them, converting all SVGs it finds.

**Comment by kaalleen** [comment]
> We could add a new extension, "Convert old SVGs" that people should run once on all of their SVGs to apply the right stroke_method. We could make it easy by processing a folder recursively for them, converting all SVGs it finds.

Uh, I would expect a storm of complaints :D On the other hand the zigzag is really terrible. We once thought about adding a zigzag option to the satin column side. Since we slide nodes around randomly, I think it shouldn't be hard to implement now. But deprecating the zigzag will not be easy.

**Comment by tyrosinase** [comment]
> We could add a new extension, "Convert old SVGs" that people should run once on all of their SVGs to apply the right stroke_method. We could make it easy by processing a folder recursively for them, converting all SVGs it finds.

FWIW I'd then have to re-generate all the archive files, and re-upload them to the website and all the marketplaces, and all the blog readers and customers would have to re-download them (which, until recently, I wasn't keeping contact information for downloads so they'd have to *know* to download it).

(Probably I should just stop including the SVG in my downloads, but dangit "open source sewing" has been my blog's tagline for twenty years or more, it feels wrong.)

**Comment by kaalleen** [comment]
@tyrosinase Do you use zigzags a lot? What are your use cases?
Maybe we can check and apply old zigzag style to all strokes up from a specific width and not by the dashes?

**Comment by tyrosinase** [comment]
>  Do you use zigzags a lot? What are your use cases?

I try not to use them at all in final designs anymore. There's a lot in older designs (you can pinpoint the changeover by going through the Issues comments and finding that one where I learned that the satin/zigzag difference was more than just a UI thing; I thought rails-and-rungs mode was just for when you needed variable width). I still use them now and then as literal zigzags (large width, large stitch spacing, either for tackdowns that don't shred vinyl or for decorative use), but not too often (and the tackdown ones don't really care if they're sawtooth or whatever). So that's not completely catastrophic; I currently have all my oldest designs offline just because I'm gonna roll them up into a "heyyyy, don't assuming my paid designs are as undocumented as these freebies are" bundle anyway.

But I use zigzags a lot, and I mean a *lot* lot, in the testing process, when I'm adjusting shapes around. Satin is a *very* lossy format. Losing the ability to tweak a single bezier and have it keep the same width would make me cry. (I'm not clear if that's what's proposed here; I've only been half-following along because I'm still mid-move. But when you said "storm of complaints" I went "oh, kaalleen is talking about me, I should read that" :laughing:)

**Comment by lexelby** [comment]
Okay, so making people convert all their old SVGs is no good.  I don't think there's a way we could make the default running stitch without breaking old SVGs...

@tyrosinase Don't worry, we're not talking about _eliminating_ simple satin, just seeing if we can make it not be the default for a line.

**Comment by kaalleen** [comment]
We could actually add a version tag to the document. Then we know exactly what to do with the old stuff.

**Comment by kaalleen** [comment]
Hehe, yes, I can't say I didn't have you in my mind :D
But it is so good to have someone take care that we don't break anything for everyone ;)

**Comment by lexelby** [comment]
> We could actually add a version tag to the document. Then we know exactly what to do with the old stuff.

I was just running over to my computer to say that :D 

Version attribute in the SVG tag or something.  If the version attribute isn't there, go tag all existing paths with a hidden attribute that says they're legacy paths, and then set the version tag.  Then if a path has the legacy tag, we treat it as it is now, otherwise we get to use the new rules.  @tyrosinase's designs all keep working the way they're supposed to.  I like it!

**Comment by tyrosinase** [comment]
> We could actually add a version tag to the document. Then we know exactly what to do with the old stuff.

I've actually noodled around with doing this somewhere, and tagging them all in the blog along with a "here's what you have to change to make this work with current Ink/Stitch." Like in some of my oldest designs, if you re-export them, you'll get a 12MM filled circle where your snaps will go, because I just plopped circles down as markers knowing that Ink/Stitch couldn't stitch them until they were converted to paths. I mean, why would that ever change in the future? :shrug: 

Also if it ever suddenly starts recognizing text and using the corresponding Lettering, I'm in trouble because I write the sizes of fabric/vinyl you need to cut for in-the-hoop designs on the corresponding layers. (At least there I don't usually do it on public-facing designs, because then it goes in the instruction PDF instead.)

**Comment by lexelby** [comment]
> Also if it ever suddenly starts recognizing text and using the corresponding Lettering, I'm in trouble because I write the sizes of fabric/vinyl you need to cut for in-the-hoop designs on the corresponding layers. (At least there I don't usually do it on public-facing designs, because then it goes in the instruction PDF instead.)

Yeah, I would be in trouble too :)

If we do this version tag system, we can handle old text properly too.

Also I realized that there is of course no need for a legacy attribute on paths, we can just go set the stroke method to reflect how Ink/Stitch previously behaved and be done with it.

**Comment by lexelby** [comment]
Okay, but how about this situation:

1. User creates a brand new SVG
2. User adds a path
3. User runs some Ink/Stitch extension
4. Ink/Stitch sees that there is no version tag and thinks it is supposed to treat that path as a legacy path, so it becomes simple satin

Am I missing a clever trick we could do to deal with that?

**Comment by lexelby** [comment]
Best I can come up with: if there is any Ink/Stitch attribute of any kind anywhere in the entire document, then we know it's an old SVG.  If not, it's a new one.

**Comment by kaalleen** [comment]
Yes, right, this makes it more complicated again, I didn'#t think about that. But I think we can live with your solution?

**Comment by lexelby** [comment]
Yup, I think so.  It does add complexity, but you've done a good job of keeping the complex legacy stuff in one place, and I'm happy with that.  Let's do it! (Maybe in another PR)

I guess we may want to call the next release a new major version of Ink/Stitch :)

**Comment by lexelby** [comment]
> Sounds good. Do you think this is ready to be merged?

I don't think so... I feel like if we're going to go with the above plan, we shouldn't be adding or removing dashes, right?  Since we're going to be ignoring dashes in the future?

**Comment by kaalleen** [comment]
Oh, I thought this should work in itself and the other ideas are for an other pull request. I can take out the dash modifications though.

**Comment by lexelby** [comment]
I think I had been thinking along those lines but then I realized it probably wouldn't work to do that.  Sorry for being confusing!

**Comment by kaalleen** [comment]
No problem :) I already removed the dash modifications. Are we good to go now?

**Comment by kaalleen** [comment]
I rebased this branch so it can use the updated inkex version.

File: lib/extensions/zigzag_line_to_satin.py [hunk 1]
```diff
@@ -50,8 +50,6 @@ def effect(self):
 
             element.node.set('d', " ".join(d))
             element.set_param('satin_column', True)
-            # remove dashes
-            element.update_dash(False)
```
Comment (lexelby) [inline lib/extensions/zigzag_line_to_satin.py:54]: I think we still need this?

**Review by lexelby — APPROVED** [review]

**Comment by lexelby** [comment]
I added a comment, but I approved so that you can merge if I'm wrong.

File: lib/extensions/zigzag_line_to_satin.py (hunk 1 above)
Comment (kaalleen) [reply lib/extensions/zigzag_line_to_satin.py:54]: At the moment yes, but not if we go forward with our idea. I'm a bit confused what to do here and what to do later :D
But the update_dash method I removed entirely when I removed the dash modifications, so either we leave it all in or just remove it at this point I suppose.

**Comment by kaalleen** [comment]
Satin columns can run on dashed lines with this branch. This means the update_dash option isn't really necessary.

So I will merge this now. Thank you for having a look at it.

**Pull request merged**